generator/camera_info_collector.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "generator/osm_element.hpp"
+#include "generator/routing_helpers.hpp"
+
+#include "routing/base/followed_polyline.hpp"
+#include "routing/routing_helpers.hpp"
+#include "routing/serialize_speed_camera.hpp"
+
+#include "indexer/data_source.hpp"
+#include "indexer/feature_data.hpp"
+#include "indexer/feature_decl.hpp"
+#include "indexer/scales.hpp"
+
+#include "coding/file_writer.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "geometry/point2d.hpp"
+#include "geometry/segment2d.hpp"
+
+#include "base/geo_object_id.hpp"
+
+#include <cstdint>
+#include <limits>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "boost/optional.hpp"
+
+namespace generator
+{
+class CamerasInfoCollector
+{
+public:
+  CamerasInfoCollector(std::string const & dataFilePath, std::string const & camerasInfoPath,
+                       std::string const & osmIdsToFeatureIdsPath);
+
+  void Serialize(FileWriter & writer) const;
+
+private:
+  struct Camera
+  {
+    static double constexpr kCoordEqualityEps = 1e-5;
+
+    Camera(m2::PointD const & center, uint8_t maxSpeed, std::vector<routing::SpeedCameraMwmPosition> && ways)
+      : m_data(center, maxSpeed, std::move(ways))
+    {
+    }
+
+    void ParseDirection()
+    {
+      // After some research we understood, that camera's direction is a random option.
+      // This fact was checked with randomly chosen cameras and google maps 3d view.
+      // The real direction and the one indicated in the OSM differed.
+      // So we left space in mwm for this feature and wait until good data appears in OSM
+    }
+
+    // Modify m_ways vector. Set for each way - id of the closest segment.
+    void FindClosestSegment(FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId);
+
+    /// \brief Try to FindClosestSegment in |m_ways|. Look to FindMyself() for more details.
+    /// \return true if we could find closest segment. And false if we should use geometry index.
+    bool FindClosestSegmentInInnerWays(FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId);
+
+    /// \brief Use when |m_ways| is empty. Try to FindClosestSegment using geometry index.
+    void FindClosestSegmentWithGeometryIndex(FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId);
+
+    // Returns empty object, if current feature - |wayId| is not the road.
+    // Otherwise returns id of segment from feature with id - |wayId|, which starts at camera's
+    // center and coefficient - where it placed at the segment.
```

Comment:
Как из документации следует, что здесь может быть только 0 или 1 в результатах?

Reply:
```cpp
// Otherwise returns id of segment from feature with id - |wayId|, which starts (or ends) at camera's
// center and coefficient - where it placed at the segment: 0.0 (or 1.0)```

okey?